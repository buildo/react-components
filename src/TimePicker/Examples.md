### Examples

#### 12h time format

```js
const [state, setState] = React.useState({});

function onChange(value) {
  setState({ value });
}

<TimePicker value={state.value} onChange={onChange} timeFormat="12h" />;
```

#### 24h time format (bounded range)

```js
const [state, setState] = React.useState({ value: { hours: 11, minutes: 23 }}});

const minTime = { hours: 8, minutes: 30 };
const maxTime = { hours: 18, minutes: 30 };

function onChange(value) {
  setState({ value });
}

<TimePicker
  value={state.value}
  minTime={minTime}
  maxTime={maxTime}
  onChange={onChange}
  timeFormat="24h"
/>;
```
