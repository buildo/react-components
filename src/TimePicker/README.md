# TimePicker

TimePicker field, used to pick a time from a dropdown

## Props
|Name|Type|Default|Description|
|----|----|-------|-----------|
| **onChange** | <code>Function</code> |  | **required**. OnChange handler. It will return an object |
| **value** | <code>Time</code> |  | *optional*. Value provided as input. Have to be passed in 24h format. E.g. { hours: 10, minutes: 30 } |
| **minTime** | <code>Time</code> | <code>{   "hours": 0,   "minutes": 0 }</code> | *optional*. Minimum value. Have to be passed in 24h format. Default [00:00] |
| **maxTime** | <code>Time</code> | <code>{   "hours": 23,   "minutes": 59 }</code> | *optional*. Maximum value. Have to be passed in 24h format. Default [23:59] |
| **placeholder** | <code>String</code> | <code>"--:--"</code> | *optional*. Field placeholder, displayed when there's no value. Default[--:--] |
| **timeFormat** | <code>enum("12h" &#124; "24h")</code> | <code>"24h"</code> | *optional*. Format in which options are displayed (12h&#124;24h) |
| **searchable** | <code>Boolean</code> | <code>true</code> | **required**. Enable the search feature |
| **id** | <code>String</code> |  | *optional*. Custom `id` for wrapper element |
| **className** | <code>String</code> |  | *optional*. Additional `className` for wrapper element |
| **style** | <code>Object</code> |  | *optional*. Inline-style overrides for wrapper element |
| **menuPosition** | <code>enum("bottom" &#124; "top")</code> | <code>"bottom"</code> | **required**. Whether the menu should open on top or bottom |
| **disabled** | <code>Boolean</code> |  | *optional*. Optionally disable the control |