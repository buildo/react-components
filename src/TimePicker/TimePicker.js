import React from 'react';
import { props, t, skinnable } from '../utils';
import cx from 'classnames';
import Dropdown from '../Dropdown/Dropdown';
import range from 'lodash/range';
import flatten from 'lodash/flatten';
import compact from 'lodash/compact';
import uniqBy from 'lodash/uniqBy';
import sortBy from 'lodash/sortBy';

export const H24 = '24h';
export const H12 = '12h';
export const inputError = {};
const symbolRegex = /[\s,\-\+\.\|\/\\]/;
const numberRegex = /^\d+$/;
const separator = ':';
const interval = 30;

const pad = (num) => num <= 9 ? `0${num}` : num;
const normalizeHoursToH24 = hour => hour + 12 !== 24 ? hour + 12 : 0;
const lteTime = (minTime, maxTime) => (
  maxTime.hours > minTime.hours || (maxTime.hours === minTime.hours && maxTime.minutes >= minTime.minutes)
);

const Integer = t.refinement(t.Number, n => n % 1 === 0, 'Integer');
const Hour = t.refinement(Integer, n => n >= 0 && n <= 23, 'Hour');
const Hour12 = t.refinement(Integer, n => n >= 1 && n <= 12, 'Hour12');
const Minute = t.refinement(Integer, n => n >= 0 && n <= 59, 'Minute');
const TimeFormat = t.enums.of([H12, H24], 'TimeFormat');
const Time = t.struct({
  hours: Hour,
  minutes: Minute
}, 'Time');

const isValidHoursInTimeFormat = (hours, timeFormat) => timeFormat === H24 ? Hour.is(hours) : Hour12.is(hours);

export const Props = t.refinement(t.struct({
  onChange: t.Function,
  value: t.maybe(Time),
  minTime: t.maybe(Time),
  maxTime: t.maybe(Time),
  placeholder: t.maybe(t.String),
  timeFormat: t.maybe(TimeFormat),
  searchable: t.maybe(t.Boolean),
  id: t.maybe(t.String),
  className: t.maybe(t.String),
  style: t.maybe(t.Object),
  menuPosition: t.enums.of(['bottom', 'top']),
  disabled: t.maybe(t.Boolean)
}), ({ value, minTime, maxTime }) => (
  lteTime(minTime, maxTime) && (!value || (lteTime(value, maxTime) && lteTime(minTime, value)))
), 'Props');

const formatTime24 = ({ hours, minutes }) => `${pad(hours)}:${pad(minutes)}`;
const formatTime12 = ({ hours, minutes }) => {
  if (hours === 0 || (hours > 12 && hours % 12 === 0)) {
    return `12${separator}${pad(minutes)} am`;
  } else if (hours === 12) {
    return `12${separator}${pad(minutes)} pm`;
  } else if (hours >= 13) {
    return `${pad(hours % 12)}${separator}${pad(minutes)} pm`;
  } else {
    return `${pad(hours)}${separator}${pad(minutes)} am`;
  }
};
const formatTime = ({ hours, minutes, timeFormat }) => (
  timeFormat === H12 ? formatTime12({ hours, minutes }) : formatTime24({ hours, minutes })
);

export const toOption = time => ({
  value: formatTime24(time),
  label: formatTime(time)
});

const cleanSeparator = str => str.replace(symbolRegex, separator);
const maybeAddTrailingZero = str => {
  if (str.length === 4) {
    return `${str}0`;
  } else {
    return str;
  }
};
const maybeInsertSeparator = str => {
  if (numberRegex.test(str)) {
    if (str.length === 3) {
      const position = str.length - 1;
      const strWithColon = `${str.substr(0, position)}${separator}${str.substr(position)}`;
      return strWithColon;
    } else if (str.length === 4) {
      const position = str.length - 2;
      const strWithColon = `${str.substr(0, position)}${separator}${str.substr(position)}`;
      return strWithColon;
    }
  }
  return str;
};

export const parseInTimeFormat = (inputStr, timeFormat) => {
  if (inputStr === '') {
    return { originalInput: inputStr };
  }
  const cleanedInput = maybeAddTrailingZero(maybeInsertSeparator(cleanSeparator(inputStr)));
  const inputArray = cleanedInput.split(separator);
  const [_hours, _minutes] = inputArray;
  const hours = numberRegex.test(_hours) ? parseInt(_hours, 10) : '';
  const minutes = !_minutes || numberRegex.test(_minutes) ? parseInt(_minutes, 10) : '';
  const validHours = isValidHoursInTimeFormat(hours, timeFormat);
  const validMinutes = Minute.is(minutes);

  if (validHours && validMinutes) {
    return { hours, minutes, originalInput: cleanedInput };
  } else if (validHours) {
    return { originalInput: cleanedInput };
  } else {
    return inputError;
  }
};

export const createTimeList = ({ hours, minutes }, timeFormat) => {
  if (!isValidHoursInTimeFormat(hours, timeFormat) || !Minute.is(minutes)) {
    const hoursList = range(0, 24);
    const minutesList = range(0, 60, interval);
    return flatten(hoursList.map(hours => minutesList.map(minutes => ({ hours, minutes, timeFormat }))));
  } else {
    return timeFormat !== H24 ? [
      { hours, minutes, timeFormat },
      { hours: normalizeHoursToH24(hours), minutes, timeFormat }
    ] : [
      { hours, minutes, timeFormat }
    ];
  }
};

const startsWith = (time, originalInput) => {
  const timeStr = formatTime(time);
  return timeStr.substr(0, originalInput.length) === originalInput;
};

export const filterTime = ({ originalInput, minTime, maxTime }) => time => {
  return lteTime(minTime, time) && lteTime(time, maxTime) && startsWith(time, originalInput);
};

export const makeOptions = ({ minTime, maxTime, timeFormat, userValue }, inputValue) => {
  const time = parseInTimeFormat(inputValue, timeFormat);
  const selectedValue = (userValue && userValue !== inputError) ? { ...userValue, timeFormat } : userValue;
  const timeList = (time === inputError ? [] : createTimeList(time, timeFormat)).concat(compact([selectedValue]));
  const filteredTimeList = timeList.filter(filterTime({
    originalInput: time.originalInput || '', minTime, maxTime
  }));
  return uniqBy(sortBy(filteredTimeList.map(toOption), 'value'), 'value');
};


/**
 * TimePicker field, used to pick a time from a dropdown
 * @param onChange - onChange handler. It will return an object
 * @param value - value provided as input. Have to be passed in 24h format. E.g. { hours: 10, minutes: 30 }
 * @param minTime - minimum value. Have to be passed in 24h format. Default [00:00]
 * @param maxTime - maximum value. Have to be passed in 24h format. Default [23:59]
 * @param placeholder - field placeholder, displayed when there's no value. Default[--:--]
 * @param timeFormat - format in which options are displayed (12h|24h)
 * @param searchable - enable the search feature
 * @param menuPosition - whether the menu should open on top or bottom
 * @param disabled - optionally disable the control
 */
@skinnable()
@props(Props)
export default class TimePicker extends React.Component {

  static defaultProps = {
    placeholder: `--${separator}--`,
    timeFormat: H24,
    minTime: { hours: 0, minutes: 0 },
    maxTime: { hours: 23, minutes: 59 },
    searchable: true,
    menuPosition: 'bottom'
  }

  state = { inputValue: '' }

  _onChange = (value) => {
    if (value) {
      // interface with component user is always in H24
      const time = parseInTimeFormat(value, H24);
      this.props.onChange(time);
    } else {
      this.props.onChange();
    }
  };

  updateInputValue = (inputValue) => this.setState({ inputValue });

  getLocals() {
    const {
      className,
      minTime, maxTime, timeFormat,
      value: userValue,
      ...props
    } = this.props;

    const value = userValue ? formatTime24(userValue) : undefined;

    return {
      ...props,
      value,
      options: makeOptions({ minTime, maxTime, timeFormat, userValue }, this.state.inputValue),
      className: cx('time-picker', className),
      onChange: this._onChange,
      updateInputValue: this.updateInputValue
    };
  }

  template({ id, className, style, searchable, value, placeholder, options, onChange, updateInputValue, menuPosition, disabled }) {
    return (
      <Dropdown
        {...{ id, className, style }}
        searchable={searchable}
        value={value}
        onChange={onChange}
        options={options}
        placeholder={placeholder}
        onInputChange={updateInputValue}
        onBlur={() => this.forceUpdate()}
        menuPosition={menuPosition}
        disabled={disabled}
      />
    );
  }
}
