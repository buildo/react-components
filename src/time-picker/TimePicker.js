import React from 'react';
import { props, t, skinnable } from '../utils';
import cx from 'classnames';
import Dropdown from '../dropdown/Dropdown';
import range from 'lodash/utility/range';
import flatten from 'lodash/array/flatten';

const options = [];

export const H24 = '24h';
export const H12 = '12h';
export const inputError = {};
const symbolRegex = /[\s,\-\+\.\|\/\\]/;
const numberRegex = /^\d+$/;
const separator = ':';
const interval = 30;

const pad = (num) => num <= 9 ? `0${num}` : num;
const normalizeHoursToH24 = hour => hour + 12 !== 24 ? hour + 12 : 0;
const lteTime = (minTime, maxTime) => (
  maxTime.hours > minTime.hours || (maxTime.hours === minTime.hours && maxTime.minutes >= minTime.minutes)
);

const Integer = t.refinement(t.Number, n => n % 1 === 0, 'Integer');
const Hour = t.refinement(Integer, n => n >= 0 && n <= 23, 'Hour');
const Hour12 = t.refinement(Integer, n => n >= 1 && n <= 12, 'Hour12');
const Minute = t.refinement(Integer, n => n >= 0 && n <= 59, 'Minute');
const TimeFormat = t.enums.of([H12, H24], 'TimeFormat');
const Time = t.struct({
  hours: Hour,
  minutes: Minute
}, 'Time');

const isValidHoursInTimeFormat = (hours, timeFormat) => timeFormat === H24 ? Hour.is(hours) : Hour12.is(hours);

const Props = t.refinement(t.struct({
  /**
  * onChange handler it will return an object
  */
  onChange: t.Function,
  /**
   * Value provided as input. Have to be passed in 24h format.
   * Es: { hours: 10, minutes: 30 }
   */
  value: t.maybe(Time),
  /**
   * Minimum value. Have to be passed in 24h format. Default [00:00]
   */
  minTime: t.maybe(Time),
  /**
   * Maximum value. Have to be passed in 24h format. Default [23:59]
   */
  maxTime: t.maybe(Time),
  /**
   * Field placeholder, displayed when there's no value. Default[--:--]
   */
  placeholder: t.maybe(t.String),
  /**
   * format in which options are displayed. Default [24H].
   */
  timeFormat: t.maybe(TimeFormat),
  id: t.maybe(t.String),
  className: t.maybe(t.String),
  style: t.maybe(t.Object)
}), ({ value, minTime, maxTime }) => (
  lteTime(minTime, maxTime) && (!value || (lteTime(value, maxTime) && lteTime(minTime, value)))
), 'Props');

const formatTime24 = ({ hours, minutes }) => `${pad(hours)}:${pad(minutes)}`;
const formatTime12 = ({ hours, minutes }) => {
  if (hours === 0 || (hours > 12 && hours % 12 === 0)) {
    return `12${separator}${pad(minutes)} am`;
  } else if (hours === 12) {
    return `12${separator}${pad(minutes)} pm`;
  } else if (hours >= 13) {
    return `${pad(hours % 12)}${separator}${pad(minutes)} pm`;
  } else {
    return `${pad(hours)}${separator}${pad(minutes)} am`;
  }
};
const formatTime = ({ hours, minutes, timeFormat }) => (
  timeFormat === H12 ? formatTime12({ hours, minutes }) : formatTime24({ hours, minutes })
);

export const toOption = time => ({
  value: formatTime24(time),
  label: formatTime(time)
});

const cleanSeparator = str => str.replace(symbolRegex, separator);
const maybeAddTrailingZero = str => {
  if (str.length === 4) {
    return `${str}0`;
  } else {
    return str;
  }
};
const maybeInsertSeparator = str => {
  if (numberRegex.test(str)) {
    if (str.length === 3) {
      const position = str.length - 1;
      const strWithColon = `${str.substr(0, position)}${separator}${str.substr(position)}`;
      return strWithColon;
    } else if (str.length === 4) {
      const position = str.length - 2;
      const strWithColon = `${str.substr(0, position)}${separator}${str.substr(position)}`;
      return strWithColon;
    }
  }
  return str;
};

export const parseInTimeFormat = (inputStr, timeFormat) => {
  if (inputStr === '') {
    return { originalInput: inputStr };
  }
  const cleanedInput = maybeAddTrailingZero(maybeInsertSeparator(cleanSeparator(inputStr)));
  const inputArray = cleanedInput.split(separator);
  const [_hours, _minutes] = inputArray;
  const hours = numberRegex.test(_hours) ? parseInt(_hours, 10) : '';
  const minutes = !_minutes || numberRegex.test(_minutes) ? parseInt(_minutes, 10) : '';
  const validHours = isValidHoursInTimeFormat(hours, timeFormat);
  const validMinutes = Minute.is(minutes);

  if (validHours && validMinutes) {
    return { hours, minutes, originalInput: cleanedInput };
  } else if (validHours) {
    return { originalInput: cleanedInput };
  } else {
    return inputError;
  }
};

export const createTimeList = ({ hours, minutes }, timeFormat) => {
  if (!isValidHoursInTimeFormat(hours, timeFormat) || !Minute.is(minutes)) {
    const hoursList = range(1, 24).concat(0);
    const minutesList = range(0, 60, interval);
    return flatten(hoursList.map(hours => minutesList.map(minutes => ({ hours, minutes, timeFormat }))));
  } else {
    return timeFormat !== H24 ? [
      { hours, minutes, timeFormat },
      { hours: normalizeHoursToH24(hours), minutes, timeFormat }
    ] : [
      { hours, minutes, timeFormat }
    ];
  }
};

const startsWith = (time, originalInput) => {
  const timeStr = formatTime(time);
  return timeStr.substr(0, originalInput.length) === originalInput;
};

export const filterTime = ({ originalInput, minTime, maxTime }) => time => {
  return lteTime(minTime, time) && lteTime(time, maxTime) && startsWith(time, originalInput);
};

// we are not "filtering" options (options array is always empty and discarded by this function)
// our `filterOptions` actually generates options based on current input string and props.
// We use this as a workaround because there's not any other easy way of updating options
// without breaking input in react-select v0.6.x
// NOTE: can be fixed updating to v1.0.0
export const makeFilterOptions = ({ minTime, maxTime, timeFormat }) => (_, inputStr) => {
  const time = parseInTimeFormat(inputStr, timeFormat);
  const timeList = time === inputError ? [] : createTimeList(time, timeFormat);
  const filteredTimeList = timeList.filter(filterTime({
    originalInput: time.originalInput, minTime, maxTime
  }));
  return filteredTimeList.map(toOption);
};

@skinnable()
@props(Props)
export default class TimePicker extends React.Component {

  static defaultProps = {
    placeholder: `--${separator}--`,
    timeFormat: H24,
    minTime: { hours: 0, minutes: 0 },
    maxTime: { hours: 23, minutes: 59 }
  }

  _onChange = (value) => {
    if (value) {
      // interface with component user is always in H24
      const time = parseInTimeFormat(value, H24);
      this.props.onChange(time);
    } else {
      this.props.onChange();
    }
  };

  getLocals() {
    const {
      timeFormat,
      className,
      value: userValue,
      ...props
    } = this.props;

    const value = userValue ? toOption({
      ...userValue, timeFormat
    }) : undefined;

    return {
      ...props,
      value,
      options,
      className: cx('time-picker', className),
      onChange: this._onChange,
      filterOptions: makeFilterOptions(this.props)
    };
  }

  template({ id, className, style, value, placeholder, options, onChange, filterOptions }) {
    return (
      <Dropdown
        {...{ id, className, style }}
        value={value}
        onChange={onChange}
        options={options}
        placeholder={placeholder}
        filterOptions={filterOptions}
        onBlur={() => this.forceUpdate()}
      />
    );
  }
}
