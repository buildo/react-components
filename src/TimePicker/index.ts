import { TimePicker } from './TimePicker';
export default TimePicker;
