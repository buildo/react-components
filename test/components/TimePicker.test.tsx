import * as React from 'react';
import { mount } from 'enzyme';
import { shallow } from 'enzyme';

import TimePicker from '../../src/TimePicker';
import {
  parseInTimeFormat,
  H24,
  H12,
  toOption,
  filterTime,
  createTimeList,
  makeOptions,
  inputError
} from '../../src/TimePicker/TimePicker';

let consoleError: jest.SpyInstance<{}>;

beforeAll(() => {
  consoleError = jest.spyOn(console, 'error');
});

afterEach(() => {
  expect(consoleError).not.toHaveBeenCalled();
});

const exampleProps: TimePicker.Props = {
  id: '12345',
  className: 'fancy-class-name',
  placeholder: '--:--',
  timeFormat: '24h',
  style: { margin: 10, position: 'relative' },
  value: { hours: 6, minutes: 30 },
  minTime: { hours: 5, minutes: 10 },
  maxTime: { hours: 8, minutes: 20 },
  onChange: () => {}
};

describe('TimePicker', () => {
  it('renders correctly', () => {
    const component = mount(<TimePicker {...exampleProps} />);
    expect(component.html()).toMatchSnapshot();
  });

  describe('props', () => {
    it('computes className', () => {
      const component = shallow(<TimePicker {...exampleProps} />);
      expect(component.hasClass('time-picker')).toBe(true);
      expect(component.hasClass('fancy-class-name')).toBe(true);
    });

    it('passes value properly formatted', () => {
      const component = shallow(
        <TimePicker value={{ hours: 15, minutes: 33 }} onChange={() => {}} />
      );
      const value = component.prop('value');
      expect(value.label).toBe('15:33');
    });

    it('passes an undefined value when value prop is undefined', () => {
      const component = shallow(<TimePicker onChange={() => {}} />);
      const value = component.prop('value');
      expect(value).toBeUndefined();
    });
  });

  describe('parseInTimeFormat', () => {
    it('parses input and returns a valid 24h time', () => {
      const input = '13:59';
      const parsedInput = parseInTimeFormat(input, H24);
      expect(parsedInput.originalInput).toBe(input);
      expect(parsedInput.hours).toBe(13);
      expect(parsedInput.minutes).toBe(59);
    });

    it('parses input and returns a valid 12h time', () => {
      const input = '06:30';
      const parsedInput = parseInTimeFormat(input, H12);
      expect(parsedInput.originalInput).toBe(input);
      expect(parsedInput.hours).toBe(6);
      expect(parsedInput.minutes).toBe(30);
    });

    it('parses input and returns an error for invalid 24h time', () => {
      const input = '36:30';
      const parsedInput = parseInTimeFormat(input, H24);
      expect(parsedInput).toBe(inputError);
    });

    it('parses input and returns an error for invalid 12h time', () => {
      const input = '18:30';
      const parsedInput = parseInTimeFormat(input, H12);
      expect(parsedInput).toBe(inputError);
    });

    it('parses empty input', () => {
      const input = '';
      const parsedInput = parseInTimeFormat(input, H24);
      expect(parsedInput.originalInput).toBe(input);
      expect(parsedInput.hours).toBeUndefined();
      expect(parsedInput.minutes).toBeUndefined();
    });

    it('parses input when minutes are not set', () => {
      const input = '10:';
      const parsedInput = parseInTimeFormat(input, H24);
      expect(parsedInput.originalInput).toBe(input);
      expect(parsedInput.hours).toBeUndefined();
      expect(parsedInput.minutes).toBeUndefined();
    });

    it('parses input without separator and return a valid time', () => {
      const input = '1231';
      const parsedInput = parseInTimeFormat(input, H24);
      expect(parsedInput.hours).toBe(12);
      expect(parsedInput.minutes).toBe(31);
      expect(parsedInput.originalInput).toBe('12:31');
    });

    it('parses input and return error if string is not well formed', () => {
      const input = '1a:aa';
      const parsedInput = parseInTimeFormat(input, H24);
      expect(parsedInput).toBe(inputError);
    });
  });

  describe('createTimeList', () => {
    it('returns all the available times', () => {
      const time = {};
      const timeList24 = createTimeList(time, H24);
      expect(timeList24.length).toBe(48);
      expect(timeList24[0].hours).toBe(0);
      expect(timeList24[0].minutes).toBe(0);
      expect(timeList24[0].timeFormat).toBe('24h');
      expect(timeList24[26].hours).toBe(13);
      expect(timeList24[26].minutes).toBe(0);
      expect(timeList24[27].hours).toBe(13);
      expect(timeList24[27].minutes).toBe(30);
      const timeList12 = createTimeList(time, H12);
      expect(timeList24).not.toEqual(timeList12);
      expect(timeList12[0].timeFormat).toBe('12h');
    });

    it('returns an array of a single element for 24h', () => {
      const time = { hours: 10, minutes: 31 };
      const timeList = createTimeList(time, H24);
      expect(timeList.length).toBe(1);
      expect(timeList[0]).toEqual({ hours: 10, minutes: 31, timeFormat: H24 });
    });

    it('returns an array of 2 elements for 12h', () => {
      const time = { hours: 10, minutes: 31 };
      const timeList = createTimeList(time, H12);
      expect(timeList.length).toBe(2);
      expect(timeList[0]).toEqual({ hours: 10, minutes: 31, timeFormat: H12 });
      expect(timeList[1]).toEqual({ hours: 22, minutes: 31, timeFormat: H12 });
    });
  });

  describe('filterTime', () => {
    it('filters based on minTime/maxTime ', () => {
      const minTime = { hours: 8, minutes: 0 };
      const maxTime = { hours: 9, minutes: 0 };
      const timeList = [
        { hours: 7, minutes: 22 },
        { hours: 8, minutes: 33 },
        { hours: 9, minutes: 44 }
      ];
      const timeListFilterd = timeList.filter(filterTime({ originalInput: '', minTime, maxTime }));
      expect(timeListFilterd.length).toBe(1);
      expect(timeListFilterd[0]).toEqual({ hours: 8, minutes: 33 });
    });

    it('filters based on original input', () => {
      const timeList = [
        { hours: 7, minutes: 22 },
        { hours: 8, minutes: 33 },
        { hours: 9, minutes: 44 }
      ];
      const timeListFilterd = timeList.filter(
        filterTime({
          originalInput: '08',
          minTime: TimePicker.defaultProps.minTime,
          maxTime: TimePicker.defaultProps.maxTime
        })
      );
      expect(timeListFilterd.length).toBe(1);
      expect(timeListFilterd[0]).toEqual({ hours: 8, minutes: 33 });
    });
  });

  describe('toOption', () => {
    it('formats into a valid option for 24H time', () => {
      const time = { hours: 10, minutes: 20 };
      const formattedTime = toOption(time, H24);
      expect(formattedTime.label).toBe('10:20');
    });

    describe('12H time', () => {
      it('works 1', () => {
        const time = { hours: 13, minutes: 20, timeFormat: H12 };
        const formattedTime = toOption(time);
        expect(formattedTime.label).toBe('01:20 pm');
      });

      it('works 2', () => {
        const time = { hours: 10, minutes: 20, timeFormat: H12 };
        const formattedTime = toOption(time);
        expect(formattedTime.label).toBe('10:20 am');
      });

      it('works 3', () => {
        const time = { hours: 12, minutes: 34, timeFormat: H12 };
        const formattedTime = toOption(time);
        expect(formattedTime.label).toBe('12:34 pm');
      });

      it('works 4', () => {
        const time = { hours: 0, minutes: 34, timeFormat: H12 };
        const formattedTime = toOption(time);
        expect(formattedTime.label).toBe('12:34 am');
      });

      it('works 5', () => {
        const time = { hours: 24, minutes: 34, timeFormat: H12 };
        const formattedTime = toOption(time);
        expect(formattedTime.label).toBe('12:34 am');
      });
    });
  });

  describe('makeOptions', () => {
    it('computes options based on min/max for 24H time', () => {
      const inputStr = '';
      const results = makeOptions(
        {
          timeFormat: H24,
          minTime: { hours: 5, minutes: 10 },
          maxTime: { hours: 7, minutes: 20 }
        },
        inputStr
      );
      expect(results.length).toBe(4);
      expect(results[0]).toEqual({
        label: '05:30',
        value: '05:30',
        time: { hours: 5, minutes: 30, timeFormat: '24h' }
      });
      expect(results[3]).toEqual({
        label: '07:00',
        value: '07:00',
        time: { hours: 7, minutes: 0, timeFormat: '24h' }
      });
    });

    it('computes options based on min/max for 12H time', () => {
      const inputStr = '';
      const results = makeOptions(
        {
          timeFormat: H12,
          minTime: { hours: 5, minutes: 10 },
          maxTime: { hours: 7, minutes: 20 }
        },
        inputStr
      );
      expect(results.length).toBe(4);
      expect(results[0]).toEqual({
        label: '05:30 am',
        value: '05:30',
        time: { hours: 5, minutes: 30, timeFormat: '12h' }
      });
      expect(results[3]).toEqual({
        label: '07:00 am',
        value: '07:00',
        time: { hours: 7, minutes: 0, timeFormat: '12h' }
      });
    });
  });
});
