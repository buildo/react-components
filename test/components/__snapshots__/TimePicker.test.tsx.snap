// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker renders correctly 1`] = `"<div class=\\"css-1pcexqc-container dropdown is-medium is-searchable time-picker fancy-class-name\\" id=\\"12345\\"><div class=\\"css-bg1rzq-control dropdown__control\\"><div class=\\"css-1hwfws3 dropdown__value-container dropdown__value-container--has-value\\"><div class=\\"css-dvua67-singleValue dropdown__single-value\\">06:30</div><div class=\\"css-1l95nvm\\"><div class=\\"dropdown__input\\" style=\\"display: inline-block;\\"><input autocapitalize=\\"none\\" autocomplete=\\"off\\" autocorrect=\\"off\\" id=\\"react-select-2-input\\" spellcheck=\\"false\\" tabindex=\\"0\\" type=\\"text\\" aria-autocomplete=\\"list\\" style=\\"box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;\\" value=\\"\\"><div style=\\"position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;\\"></div></div></div></div><div class=\\"css-1wy0on6 dropdown__indicators\\"><div aria-hidden=\\"true\\" class=\\"css-16pqwjk-indicatorContainer dropdown__indicator dropdown__dropdown-indicator\\"><svg height=\\"20\\" width=\\"20\\" viewBox=\\"0 0 20 20\\" aria-hidden=\\"true\\" focusable=\\"false\\" class=\\"css-19bqh2r\\"><path d=\\"M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z\\"></path></svg></div></div></div></div>"`;
